clients/bot_client.go

```diff
@@ -142,3 +150,136 @@ func (botClient *BotClient) Sync() {
+func (botClient *BotClient) SubmitDecimalSAS(otherUser id.UserID, otherDevice id.DeviceID, sas crypto.DecimalSASData) {
+	key := otherUser.String() + ":" + otherDevice.String()
+	sasChan, loaded := botClient.verificationSAS.LoadOrStore(key, make(chan crypto.DecimalSASData))
+	go func() {
+		if !loaded {
+			// if we created the chan, delete it after the timeout duration
+			defer botClient.verificationSAS.Delete(key)
+		}
+		// insert to channel in goroutine to avoid blocking if we are not expecting a SAS for this user/device right now
+		select {
+		case sasChan.(chan crypto.DecimalSASData) <- crypto.DecimalSASData(sas):
+		case <-time.After(botClient.olmMachine.DefaultSASTimeout):
+			log.Warnf("Timed out while trying to send SAS for device %v", otherDevice)
+		}
+	}()
+}
+
+// VerificationMethods returns the supported SAS verification methods.
+// As a bot we only support decimal as it's easier to understand.
+func (botClient *BotClient) VerificationMethods() []crypto.VerificationMethod {
+	return []crypto.VerificationMethod{
+		crypto.VerificationMethodDecimal{},
+	}
+}
+
+// OnCancel is called when a SAS verification is canceled.
+func (botClient *BotClient) OnCancel(cancelledByUs bool, reason string, reasonCode event.VerificationCancelCode) {
+}
+
+// OnSuccess is called when a SAS verification is successful.
+func (botClient *BotClient) OnSuccess() {
+}
+
+// InvalidateRoomSession invalidates the outbound group session for the given room.
+func (botClient *BotClient) InvalidateRoomSession(roomID id.RoomID) (id.SessionID, error) {
+	outbound, err := botClient.olmMachine.CryptoStore.GetOutboundGroupSession(roomID)
+	if err != nil {
+		return "", err
+	}
+	if outbound == nil {
+		return "", errors.New("No group session found for this room")
+	}
+	return outbound.ID(), botClient.olmMachine.CryptoStore.RemoveOutboundGroupSession(roomID)
+}
+
+// StartSASVerification starts a new SAS verification with the given user and device ID and returns the transaction ID if successful.
+func (botClient *BotClient) StartSASVerification(userID id.UserID, deviceID id.DeviceID) (string, error) {
+	device, err := botClient.olmMachine.GetOrFetchDevice(userID, deviceID)
+	if err != nil {
+		return "", err
+	}
+	return botClient.olmMachine.NewSimpleSASVerificationWith(device, botClient)
+}
+
+// SendRoomKeyRequest sends a room key request to another device.
+func (botClient *BotClient) SendRoomKeyRequest(userID id.UserID, deviceID id.DeviceID, roomID id.RoomID,
+	senderKey id.SenderKey, sessionID id.SessionID, timeout time.Duration) (chan bool, error) {
+
+	ctx, _ := context.WithTimeout(context.Background(), timeout)
+	return botClient.olmMachine.RequestRoomKey(ctx, userID, deviceID, roomID, senderKey, sessionID)
+}
+
+// ForwardRoomKeyToDevice sends a room key to another device.
+func (botClient *BotClient) ForwardRoomKeyToDevice(userID id.UserID, deviceID id.DeviceID, roomID id.RoomID, senderKey id.SenderKey,
```

Comment:
When is this function called? What controls the user/device?

Reply:
I needed this in cryptotest for one of the functionalities (`!forward_me_room_key`). So whoever calls that cmd receives the room key they asked for that room (the one where cryptotest runs).